obspy/clients/fdsn/mass_downloader/download_helpers.py

```diff
@@ -0,0 +1,1140 @@
+
+    def __len__(self):
+        return len(self.stations)
+
+    def prepare_mseed_download(self):
+        """
+        Prepare each Station for the MiniSEED downloading stage.
+
+        This will distribute filenames and identify files that require
+        downloading.
+        """
+        for station in self.stations.values():
+            station.prepare_mseed_download(mseed_storage=self.mseed_storage)
+
+    def filter_stations_based_on_minimum_distance(
+            self, existing_client_dl_helpers):
+        """
+        Removes stations until all stations have a certain minimum distance to
+        each other.
+
+        Returns the rejected stations which is mainly useful for testing.
+
+        :param existing_client_dl_helpers: Instances of already existing
+            client download helpers.
+        :type existing_client_dl_helpers: list of
+            :class:`~.ClientDownloadHelper`
+        """
+        if not self.restrictions.minimum_interstation_distance_in_m:
+            # No rejected stations.
+            return []
+
+        # Create a sorted copy that will be used in the following. Make it
+        # more deterministic by sorting the stations based on the id.
+        stations = copy.copy(list(self.stations.values()))
+        stations = sorted(stations, key=lambda x: (x.network, x.station))
+
+        existing_stations = []
+        for dlh in existing_client_dl_helpers:
+            existing_stations.extend(list(dlh.stations.values()))
+
+        remaining_stations = []
+        rejected_stations = []
+
+        # There are essentially two possibilities. If no station exists yet,
+        # it will choose the largest subset of stations satisfying the
+        # minimum inter-station distance constraint.
+        if not existing_stations and \
+                not any([_i.has_existing_time_intervals for _i in stations]):
+            # Build k-d-tree and query for the neighbours of each point within
+            # the minimum distance.
+            kd_tree = utils.SphericalNearestNeighbour(stations)
+            nns = kd_tree.query_pairs(
+                self.restrictions.minimum_interstation_distance_in_m)
+
+            indexes_to_remove = []
+            # Keep removing the station with the most pairs until no pairs are
+            # left.
+            while nns:
+                most_common = collections.Counter(
+                    itertools.chain.from_iterable(nns)).most_common()[0][0]
+                indexes_to_remove.append(most_common)
+                nns = list(itertools.filterfalse(
+                    lambda x: most_common in x, nns))
+
+            # Remove these indices this results in a set of stations we wish to
+            # keep.
+            remaining_stations.extend(
+                set([_i[1] for _i in itertools.filterfalse(
+                    lambda x: x[0] in indexes_to_remove,
+                    enumerate(stations))]))
```

Comment:
Is `itertools` really necessary? This is equivalent to `[_i[1] for _i in enumerate(stations) if x[0] not in indexes_to_remove]`, isn't it (which seems simpler to me)?